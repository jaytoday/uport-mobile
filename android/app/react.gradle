/*
 * Copyright (C) 2018 ConsenSys AG
 *
 * This file is part of uPort Mobile App.
 *
 * uPort Mobile App is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version 3
 * of the License, or (at your option) any later version.
 *
 * uPort Mobile App is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with uPort Mobile App.
 * If not, see <http://www.gnu.org/licenses/>.
 *
 */

import org.apache.tools.ant.taskdefs.condition.Os

def config = project.hasProperty("react") ? project.react : [];

def bundleAssetName = config.bundleAssetName ?: "index.android.bundle"
def entryFile = config.entryFile ?: "index.js"

// because elvis operator
def elvisFile(thing) {
    return thing ? file(thing) : null;
}

def reactRoot = elvisFile(config.root) ?: file("../../")
def inputExcludes = config.inputExcludes ?: ["android/**", "ios/**"]

void runBefore(String dependentTaskName, Task task) {
    Task dependentTask = tasks.findByPath(dependentTaskName);
    if (dependentTask != null) {
        dependentTask.dependsOn task
    }
}

gradle.projectsEvaluated {
    // Grab all build types and product flavors
    def buildTypes = android.buildTypes.collect { type -> type.name }
    def productFlavors = android.productFlavors.collect { flavor -> flavor.name }

    // When no product flavors defined, use empty
    if (!productFlavors) productFlavors.add('')

    productFlavors.each { productFlavorName ->
        buildTypes.each { buildTypeName ->
            // Create variant and target names
            def targetName = "${productFlavorName.capitalize()}${buildTypeName.capitalize()}"
            def targetPath = productFlavorName ?
                    "${productFlavorName}/${buildTypeName}" :
                    "${buildTypeName}"

            // React js bundle directories
            def jsBundleDirConfigName = "jsBundleDir${targetName}"
            def jsBundleDir = elvisFile(config."$jsBundleDirConfigName") ?:
                    file("$buildDir/intermediates/assets/${targetPath}")

            def resourcesDirConfigName = "resourcesDir${targetName}"
            def resourcesDir = elvisFile(config."${resourcesDirConfigName}") ?:
                    file("$buildDir/intermediates/res/merged/${targetPath}")
            def jsBundleFile = file("$jsBundleDir/$bundleAssetName")

            // Bundle task name for variant
            def bundleJsAndAssetsTaskName = "bundle${targetName}JsAndAssets"

            def currentBundleTask = tasks.create(
                    name: bundleJsAndAssetsTaskName,
                    type: Exec) {
                group = "react"
                description = "bundle JS and assets for ${targetName}."

                // Create dirs if they are not there (e.g. the "clean" task just ran)
                doFirst {
                    jsBundleDir.mkdirs()
                    resourcesDir.mkdirs()
                }

                // Set up inputs and outputs so gradle can cache the result
                inputs.files fileTree(dir: reactRoot, excludes: inputExcludes)
                outputs.dir jsBundleDir
                outputs.dir resourcesDir

                // Set up the call to the react-native cli
                workingDir reactRoot

                // Set up dev mode
                def devEnabled = !targetName.toLowerCase().contains("release")
                if (Os.isFamily(Os.FAMILY_WINDOWS)) {
                    commandLine "cmd", "/c", "react-native", "bundle", "--platform", "android", "--dev", "${devEnabled}",
                            "--entry-file", entryFile, "--bundle-output", jsBundleFile, "--assets-dest", resourcesDir
                } else {
                    commandLine "react-native", "bundle", "--platform", "android", "--dev", "${devEnabled}",
                            "--entry-file", entryFile, "--bundle-output", jsBundleFile, "--assets-dest", resourcesDir
                }

                enabled config."bundleIn${targetName}" ||
                    config."bundleIn${buildTypeName.capitalize()}" ?:
                            targetName.toLowerCase().contains("release")
            }

            // Hook bundle${productFlavor}${buildType}JsAndAssets into the android build process
            currentBundleTask.dependsOn("merge${targetName}Resources")
            currentBundleTask.dependsOn("merge${targetName}Assets")

            runBefore("processArmeabi-v7a${targetName}Resources", currentBundleTask)
            runBefore("processX86${targetName}Resources", currentBundleTask)
            runBefore("processUniversal${targetName}Resources", currentBundleTask)
            runBefore("process${targetName}Resources", currentBundleTask)
        }
    }
}
