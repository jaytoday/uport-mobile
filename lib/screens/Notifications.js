import React from 'react'
import { StyleSheet, Text, View, VirtualizedList, Platform } from 'react-native'
import { Navigation } from 'react-native-navigation'
import { connect } from 'react-redux'
import { count, toJs, get } from 'mori'
import { colors } from 'uPortMobile/lib/styles/globalStyles'
// Selectors
import { notifications, notificationCount } from 'uPortMobile/lib/selectors/activities'
import { externalProfile } from 'uPortMobile/lib/selectors/requests'
import { requestedClaims } from 'uPortMobile/lib/selectors/attestations'

// Actions
import { authorizeRequest, cancelRequest, selectRequest } from 'uPortMobile/lib/actions/requestActions'
import { openActivity } from 'uPortMobile/lib/actions/uportActions'

// Components
import TransactionNotification from './notifications/variants/TransactionNotification'
import DisclosureNotification from './notifications/variants/DisclosureNotification'
import VerificationSignNotification from './notifications/variants/VerificationSignNotification'
import ConnectionNotification from './notifications/variants/ConnectionNotification'
import RecoveryNotification from './notifications/variants/RecoveryNotification'
import AttestationNotification from './notifications/variants/AttestationNotification'
import NetworkNotification from './notifications/variants/NetworkNotification'
import InformationNotification from './notifications/variants/InformationNotification'
import InfoNotification from './notifications/variants/InfoNotification'
import SignTypedDataNotification from './notifications/variants/SignTypedDataNotification'

import { Icon } from '@kancha'

export class Notifications extends React.Component {
  constructor(props) {
    super(props)
    this.getNotificationType = this.getNotificationType.bind(this)
    this.renderItem = this.renderItem.bind(this)
  }

  componentDidMount() {
    this.updateInAppNotificationBadge(this.props.unreadNoticationCount)
  }

  getNotificationType(item, issuer, selectRequest) {
    const activity = toJs(item)

    switch (activity.type) {
      case 'sign':
        return (
          <TransactionNotification
            activity={activity}
            issuer={issuer}
            selectRequest={selectRequest}
            cancel={this.props.cancelRequest}
            authorize={this.props.authorizeRequest}
          />
        )
      case 'disclosure':
        return (
          <DisclosureNotification
            activity={activity}
            issuer={issuer}
            selectRequest={selectRequest}
            cancel={this.props.cancelRequest}
            authorize={this.props.authorizeRequest}
            requested={this.props.requested}
          />
        )
      case 'verificationSign':
        return (
          <VerificationSignNotification
            activity={activity}
            issuer={issuer}
            selectRequest={selectRequest}
            cancel={this.props.cancelRequest}
            authorize={this.props.authorizeRequest}
          />
        )
      case 'connect':
        return (
          <ConnectionNotification
            activity={activity}
            issuer={issuer}
            selectRequest={selectRequest}
            cancel={this.props.cancelRequest}
            authorize={this.props.authorizeRequest}
          />
        )
      case 'attestation':
        return (
          <AttestationNotification
            activity={activity}
            issuer={issuer}
            selectRequest={selectRequest}
            cancel={this.props.cancelRequest}
            authorize={this.props.selectRequest}
          />
        )
      case 'recover':
        return (
          <RecoveryNotification
            activity={activity}
            issuer={issuer}
            selectRequest={selectRequest}
            cancel={this.props.cancelRequest}
            authorize={this.props.authorizeRequest}
          />
        )
      case 'net':
        return (
          <NetworkNotification
            activity={activity}
            issuer={issuer}
            selectRequest={selectRequest}
            cancel={this.props.cancelRequest}
            authorize={this.props.authorizeRequest}
          />
        )
      case 'information':
        return (
          <InformationNotification
            activity={activity}
            infoType={activity.infoType}
            selectRequest={() => {}}
            cancel={this.props.cancelRequest}
            authorize={this.props.authorizeRequest}
          />
        )
      case 'infoReq':
        return (
          <InfoNotification
            activity={activity}
            issuer={issuer}
            infoType={activity.infoType}
            selectRequest={() => {}}
            cancel={this.props.cancelRequest}
            authorize={this.props.authorizeRequest}
          />
        )
      case 'eip712Sign':
        return (
          <SignTypedDataNotification
            activity={activity}
            issuer={issuer}
            selectRequest={selectRequest}
            cancel={this.props.cancelRequest}
            authorize={this.props.selectRequest}
          />
        )
    }
  }
  renderItem({ item, index }) {
    return (
      <View key={`activity-${item.id}`}>
        {this.getNotificationType(item, this.props.lookupIssuer, this.props.selectRequest)}
      </View>
    )
  }

  updateInAppNotificationBadge(count) {
    Navigation.mergeOptions(this.props.componentId, {
      bottomTab: {
        badge: (count > 0 && count.toString()) || null,
        badgeColor: colors.red,
      },
    })
  }

  componentDidUpdate(prevProps) {
    if (prevProps.unreadNoticationCount !== this.props.unreadNoticationCount) {
      this.updateInAppNotificationBadge(this.props.unreadNoticationCount)
    }
  }

  renderNoNotifications() {
    return (
      <View style={styles.emptyStateContainer}>
        <Icon name={'bell'} font={'feather'} color={colors.white226} size={100} />
        <Text style={{ fontSize: 22, color: colors.grey91, paddingTop: 20 }}>You don't have any notifications</Text>
      </View>
    )
  }

  renderNotifications() {
    return (
      <View style={styles.container}>
        <VirtualizedList
          data={this.props.notifications}
          getItem={get}
          getItemCount={count}
          renderItem={this.renderItem}
          keyExtractor={i => get(i, 'id').toString()}
          contentContainerStyle={styles.list}
        />
      </View>
    )
  }

  render() {
    return count(this.props.notifications) ? this.renderNotifications() : this.renderNoNotifications()
  }
}

const styles = StyleSheet.create({
  emptyStateContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: colors.white,
  },
  container: {
    flex: 1,
    justifyContent: 'flex-start',
    alignItems: 'stretch',
    backgroundColor: colors.white246,
  },
  list: {
    margin: 4,
    marginTop: 26,
    alignItems: 'stretch',
    justifyContent: 'flex-start',
  },
})

const mapStateToProps = state => ({
  unreadNoticationCount: notificationCount(state),
  lookupIssuer: clientId => toJs(externalProfile(state, clientId)),
  notifications: notifications(state),
  requested: activity => requestedClaims(state, activity),
})

export const mapDispatchToProps = dispatch => ({
  selectRequest: activity => dispatch(selectRequest(activity.id)),
  authorizeRequest: activity => dispatch(authorizeRequest(activity.id)),
  cancelRequest: activity => dispatch(cancelRequest(activity.id)),
  hideRequest: activity => dispatch(openActivity(activity.id)), // marking an activity as opened hides it
})

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Notifications)
