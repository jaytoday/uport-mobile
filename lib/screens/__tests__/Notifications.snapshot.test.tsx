import * as React from 'react'
import { render } from 'react-native-testing-library'
import { Notifications } from '../Notifications'

jest.mock('react-native-navigation', () => {
  return {
    Navigation: {
      mergeOptions: () => true,
    },
  }
})

describe('Notifications', () => {
  it('should render default screen', () => {
    const tree = render(<Notifications />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
