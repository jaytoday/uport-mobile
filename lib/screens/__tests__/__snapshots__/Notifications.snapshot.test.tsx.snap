// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications should render default screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": "rgba(226,226,226,1)",
          "fontSize": 100,
        },
        undefined,
        Object {
          "fontFamily": "Feather",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    ï„ž
  </Text>
  <Text
    style={
      Object {
        "color": "rgba(91,91,91,1)",
        "fontSize": 22,
        "paddingTop": 20,
      }
    }
  >
    You don't have any notifications
  </Text>
</View>
`;
